netgear/netgear.go

```diff
@@ -0,0 +1,177 @@
+package netgear
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const timeout = 10 * time.Second
+
+type Netgear struct {
+	sessionToken string
+	client       *http.Client
+
+	baseUrl  *url.URL
+	username string
+	password string
+}
+
+func NewNetgear(baseUrl, username, password string) (*Netgear, error) {
+	u, err := url.Parse(strings.TrimSpace(baseUrl))
+	if err != nil {
+		return nil, err
+	}
+	// be sure that the base path is /api/v1 using JoinPath
+	u = u.JoinPath("api", "v1")
+
+	return &Netgear{
+		client:   &http.Client{Timeout: timeout},
+		baseUrl:  u,
+		username: username,
+		password: password,
+	}, nil
+}
+
+func (n *Netgear) Login() error {
+	loginPayload := map[string]interface{}{
+		"user": map[string]string{
+			"name":     n.username,
+			"password": n.password,
+		},
+	}
+
+	payloadBytes, err := json.Marshal(loginPayload)
+	if err != nil {
+		return err
+	}
+
+	loginURL := n.baseUrl.JoinPath("login")
+	req, err := http.NewRequest(http.MethodPost, loginURL.String(), bytes.NewBuffer(payloadBytes))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := n.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
```

Comment:
> If the Body is not both read to EOF and closed, the [Client](https://pkg.go.dev/net/http#Client)'s underlying [RoundTripper](https://pkg.go.dev/net/http#RoundTripper) (typically [Transport](https://pkg.go.dev/net/http#Transport)) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

-- <https://pkg.go.dev/net/http#Client.Do>

---

Your `json.NewDecoder` might fail for unknown input or whatnot. Just make sure to always drain the response body.


```suggestion
	defer func() {
		_, _ = io.Copy(io.Discard, resp.Body)
		_ = resp.Body.Close()
	}()
```